### PR TITLE
Pack size and align into a single vtable entry

### PR DESCRIPTION
**Pull request by GnomedDev**
As-is, each vtable includes a header metadata segment containing `drop_ptr, size, align` as `usize`s. However, with a smart encoding taking advantage of the `isize::MAX` cap on type sizes, this can be shortened to `drop_ptr, layout` with hopefully very minimal runtime costs. This PR implements said encoding and decoding logic. 

Thanks to scottmcm for the packing and unpacking algorithms.

### CODE REVIEW

**Comment by saethlin** [comment]
We want to know how much this regresses binary size
@bors try @rust-timer queue

**Comment by saethlin** [comment]
ripgrep opt-full and exa opt-full got a bit bigger, and those are the only two benchmarks that build release binaries. Oof.

**Comment by GnomedDev** [comment]
I'll look into this and if I can outline some of the unpacking logic.

**Comment by GnomedDev** [comment]
Most differences, as expected, seem to be in `Box::drop` and other callers of `Layout::from_value_raw`, however I can't get the compiler to nicely only outline the unsized case and keep the sized case inlining nicely. It might be an idea to see if moving some of this logic from intrinsics to normal core functions could improve things. 

@rustbot author

File: compiler/rustc_middle/src/ty/vtable.rs [hunk 1]
```diff
@@ -10,10 +10,8 @@ use crate::ty::{self, Instance, PolyTraitRef, Ty, TyCtxt};
 pub enum VtblEntry<'tcx> {
     /// destructor of this type (used in vtable header)
     MetadataDropInPlace,
-    /// layout size of this type (used in vtable header)
-    MetadataSize,
-    /// layout align of this type (used in vtable header)
-    MetadataAlign,
+    /// size and align of this type (used in vtable header)
```
Comment (RalfJung) [inline compiler/rustc_middle/src/ty/vtable.rs:13]: (Only relevant if this proceeds to an actual proposal.)

Please explain how the encoding works.

**Comment by Dylan-DPC** [comment]
@GnomedDev any updates on this? thanks

**Comment by GnomedDev** [comment]
Nope, I'll close this as I don't have enough experience around rustc to figure out how to outline the vtable unpacking, and even then I doubt that will be good for performance.

**Pull request closed without merging**